From 9734fc84bc0be272a42eafcba8dac9792cf2592c Mon Sep 17 00:00:00 2001
From: Tommy Chen <tommyyl.chen@mediatek.com>
Date: Fri, 29 Aug 2025 15:30:33 +0800
Subject: [PATCH 3/3] GENIO: arm64: dts: mt8391: Add DVO and eDP nodes

Add DVO and eDP nodes for customers verifying the eDP panel

Signed-off-by: Tommy Chen <tommyyl.chen@mediatek.com>
---
 arch/arm64/boot/dts/mediatek/mt8189.dtsi      |  36 +++++
 .../dts/mediatek/mt8391-genio-720-evk.dts     | 136 +++++++++++++++++-
 2 files changed, 168 insertions(+), 4 deletions(-)

diff --git a/arch/arm64/boot/dts/mediatek/mt8189.dtsi b/arch/arm64/boot/dts/mediatek/mt8189.dtsi
index 030e72631aa5..cd0905b12023 100644
--- a/arch/arm64/boot/dts/mediatek/mt8189.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt8189.dtsi
@@ -40,6 +40,7 @@ aliases {
 		rdma1 = &disp_rdma1;
 		mmc0 = &mmc0;
 		mmc1 = &mmc1;
+		dvo0 = &disp_dvo0;
 	};
 
 	cpus {
@@ -442,6 +443,41 @@ pericfg_ao_clk: syscon@11036000 {
 			#clock-cells = <1>;
 		};
 
+
+		disp_dvo0: disp-dvo0@14017000 {
+			compatible = "mediatek,mt8189-edp-dvo";
+			status = "okay";
+			reg = <0 0x14017000 0 0x1000>;
+			interrupts = <GIC_SPI 325 IRQ_TYPE_LEVEL_HIGH 0>;
+			clocks = <&dispsys_config_clk CLK_MMSYS_1_DISP_DVO>,
+						<&topckgen_clk CLK_TOP_EDP_SEL>,
+						<&apmixedsys_clk CLK_APMIXED_TVDPLL2>,
+						<&topckgen_clk CLK_TOP_TCK_26M_MX9>,
+						<&dispsys_config_clk CLK_MMSYS_0_DISP_DVO>,
+						<&topckgen_clk CLK_TOP_TVDPLL2_D2>,
+						<&topckgen_clk CLK_TOP_TVDPLL2_D4>,
+						<&topckgen_clk CLK_TOP_TVDPLL2_D8>,
+						<&topckgen_clk CLK_TOP_TVDPLL2_D16>;
+			clock-names = "engine",
+						"pixel",
+						"pll",
+						"dvo_clk",
+						"hf_fdvo_clk",
+						"TVDPLL2_D2",
+						"TVDPLL2_D4",
+						"TVDPLL2_D8",
+						"TVDPLL2_D16";
+		};
+
+		edp_tx: edp-tx@11b70000 {
+			compatible = "mediatek,mt8189-edp-tx";
+			status = "okay";
+			reg = <0 0x11b70000 0 0x14000>,
+				<0 0x11e10000 0 0x1500>;
+			power-domains = <&scpsys MT8189_POWER_DOMAIN_EDP_TX_DORMANT>;
+			interrupts = <GIC_SPI 380 IRQ_TYPE_LEVEL_HIGH 0>;
+		};
+
 		uart0: serial@11001000 {
 			compatible = "mediatek,mt6577-uart";
 			reg = <0 0x11001000 0 0x1000>;
diff --git a/arch/arm64/boot/dts/mediatek/mt8391-genio-720-evk.dts b/arch/arm64/boot/dts/mediatek/mt8391-genio-720-evk.dts
index bed552432768..0fcc467dda2a 100644
--- a/arch/arm64/boot/dts/mediatek/mt8391-genio-720-evk.dts
+++ b/arch/arm64/boot/dts/mediatek/mt8391-genio-720-evk.dts
@@ -36,6 +36,37 @@ backlight_lcd0: backlight-lcd0 {
 		status = "okay";
 	};
 
+	backlight_lcd1: backlight-lcd1 {
+		compatible = "pwm-backlight";
+		pwms = <&disp_pwm1 0 500000>;
+		brightness-levels = <0 1023>;
+		num-interpolated-steps = <1023>;
+		default-brightness-level = <576>;
+		status = "okay";
+	};
+
+	edp_panel_fixed_3v3: edp-panel-fixed-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "edp_panel_3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		enable-active-high;
+		gpio = <&pio 109 GPIO_ACTIVE_HIGH>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&edp_panel_3v3_en_pins>;
+	};
+
+	edp_panel_fixed_12v: edp-panel-fixed-12v {
+		compatible = "regulator-fixed";
+		regulator-name = "edp_backlight_12v";
+		regulator-min-microvolt = <12000000>;
+		regulator-max-microvolt = <12000000>;
+		enable-active-high;
+		gpio = <&pio 110 GPIO_ACTIVE_HIGH>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&edp_panel_12v_en_pins>;
+	};
+
 	chosen {
 		stdout-path = "serial0:921600n8";
 	};
@@ -428,6 +459,11 @@ &disp_pwm0 {
 	pinctrl-0 = <&disp_pwm0_pin_default>;
 };
 
+&disp_pwm1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&disp_pwm1_pin_default>;
+};
+
 &pio {
 	uart0_pins: uart0-pins {
 		pins {
@@ -852,7 +888,7 @@ pins-cmd1-dat {
 			bias-pull-down;
 		};
 
-		pins-cmd2-dat {
+		/*pins-cmd2-dat {
 			pinmux = <PINMUX_GPIO108__FUNC_GPIO108>;
 			input-enable;
 			bias-pull-down;
@@ -868,7 +904,7 @@ pins-cmd4-dat {
 			pinmux = <PINMUX_GPIO110__FUNC_GPIO110>;
 			input-enable;
 			bias-pull-down;
-		};
+		};*/
 	};
 
 	aud_gpio_i2sout0_on: aud-gpio-i2sout0-on {
@@ -878,7 +914,7 @@ pins-cmd1-dat {
 			bias-disable;
 		};
 
-		pins-cmd2-dat {
+		/*pins-cmd2-dat {
 			pinmux = <PINMUX_GPIO108__FUNC_I2SOUT0_BCK>;
 			input-schmitt-enable;
 			bias-disable;
@@ -894,7 +930,7 @@ pins-cmd4-dat {
 			pinmux = <PINMUX_GPIO110__FUNC_I2SOUT0_DO>;
 			input-schmitt-enable;
 			bias-disable;
-		};
+		};*/
 	};
 
 	/* Audio DMIC Interface */
@@ -1037,6 +1073,98 @@ pins-cmd6-dat {
 		};
 	};
 	/* AUDIO GPIO standardization end */
+
+	edp_pin: edp-pin-default {
+		pins-cmd-dat {
+			pinmux = <PINMUX_GPIO28__FUNC_EDP_TX_HPD>;
+			bias-pull-up;
+		};
+	};
+
+	disp_pwm1_pin_default: disp-pwm1-pin-default {
+		pins1 {
+			pinmux = <PINMUX_GPIO30__FUNC_DISP_PWM1>;
+		};
+	};
+
+	edp_panel_12v_en_pins: edp-panel-12v-en-pins {
+		pins1 {
+			pinmux = <PINMUX_GPIO110__FUNC_GPIO110>;
+			output-high;
+		};
+	};
+
+	edp_panel_3v3_en_pins: edp-panel-3v3-en-pins {
+		pins1 {
+			pinmux = <PINMUX_GPIO109__FUNC_GPIO109>;
+			output-high;
+		};
+	};
+};
+
+&disp_dvo0 {
+	status = "okay";
+
+	ports {
+		port {
+			disp_dvo0_out: endpoint {
+				remote-endpoint = <&edptx_in>;
+			};
+		};
+	};
+};
+
+&edp_tx {
+	pinctrl-names = "default";
+	pinctrl-0 = <&edp_pin>;
+	max-lane-count = <4>;
+	max-linkrate-mhz = <5400>;
+	use-aux-bus;
+
+	status = "okay";
+
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			edptx_in: endpoint {
+				remote-endpoint = <&disp_dvo0_out>;
+			};
+		};
+
+		port@1 {
+			reg = <1>;
+			edp_out: endpoint {
+				data-lanes = <0 1 2 3>;
+				remote-endpoint = <&edp_panel_in>;
+			};
+		};
+	};
+
+	aux-bus {
+		panel: panel {
+			/*compatible = "innolux,hk173vb-01b";*/
+			compatible = "auo,g156han03";
+			power-supply = <&edp_panel_fixed_3v3>;
+			backlight = <&backlight_lcd1>;
+
+			status = "okay";
+
+			port {
+				edp_panel_in: endpoint {
+					remote-endpoint = <&edp_out>;
+				};
+			};
+		};
+	};
+};
+
+&backlight_lcd1 {
+	default-on;
+	power-supply = <&edp_panel_fixed_12v>;
+	enable-gpios = <&pio 108 GPIO_ACTIVE_HIGH>;
 };
 
 &mmc1 {
-- 
2.45.2

